packages/cli/src/execution-lifecycle/execution-lifecycle-hooks.ts

```diff
@@ -34,6 +34,61 @@ import {
 } from './shared/shared-hook-functions';
 import { toSaveSettings } from './to-save-settings';
 
+function mergeHookFunctions(...hookFunctions: IWorkflowExecuteHooks[]): IWorkflowExecuteHooks {
+	const result: IWorkflowExecuteHooks = {
+		nodeExecuteBefore: [],
+		nodeExecuteAfter: [],
+		workflowExecuteBefore: [],
+		workflowExecuteAfter: [],
+		sendResponse: [],
+		nodeFetchedData: [],
+	};
+	for (const hooks of hookFunctions) {
+		for (const key in hooks) {
+			if (hooks.hasOwnProperty(key)) {
+				result[key]!.push(...hooks[key]!);
+			}
+		}
+	}
+	return result;
+}
+
+function hookFunctionsWorkflowTelemetry(userId?: string): IWorkflowExecuteHooks {
+	const eventService = Container.get(EventService);
+	return {
+		workflowExecuteBefore: [
+			async function (this: WorkflowHooks): Promise<void> {
+				const { executionId, workflowData } = this;
+				eventService.emit('workflow-pre-execute', { executionId, data: workflowData });
+			},
+		],
+		workflowExecuteAfter: [
+			async function (this: WorkflowHooks, runData: IRun): Promise<void> {
```

Comment:
What's our policy on inferrable return types? Personally I find them noisy.

Reply:
quite a lot of explicit typing in this file will be removed after I refactor WorkflowHooks to be type-safe.